sentry/src/routes/campaign.rs

```diff
@@ -697,110 +697,144 @@ pub mod insert_events {
+            session.clone(),
+            events_success,
+        );
 
-                        // Log unsuccessfully spending
-                        match spending_result {
-                            Ok(()) => Some((event, earner, payout)),
-                            Err(err) => {
-                                error!(&app.logger, "Payout spending failed: {}", err; "campaign" => ?campaign_context, "event" => ?event, "earner" => ?earner, "unpaid amount" => %payout, "err" => ?err);
+        Ok(true)
+    }
 
-                                None
-                            }
-                        }
+    /// Max retries is `5` after which an error logging message will be recorded.
+    fn analytics_record_spawn(
+        pool: DbPool,
+        logger: Logger,
+        campaign_context: ChainOf<Campaign>,
+        session: Session,
+        events: Vec<(Event, Address, UnifiedNum)>,
+    ) {
+        tokio::spawn(async move {
+            let max_retries = 5;
+
+            for retry in 1..=5 {
+                let result = analytics::record(&pool, &campaign_context, &session, &events).await;
+
+                match (result, retry) {
+                    // if the recording was successful, break the loop early
+                    (Ok(_), _) => break,
+                    // if we have reached the max retries, log a message
+                    (Err(err), retry) if retry == max_retries => {
+                        error!(&logger, "Analytics recording failed after {max_retries} retries: {}", err; "campaign" => ?campaign_context, "err" => ?err, "events" => ?events, "session" => ?session);
                     }
-                    // if None, then ad was never show
-                    None => None,
+                    // If we have not reached the max_retries, just continue to the next loop
+                    _ => {}
                 }
-            };
-
-            if let Some(event_payout) = event_payout {
-                events_success.push(event_payout);
             }
-        }
-
-        // Record successfully paid out events to Analytics
-        if let Err(err) =
-            analytics::record(&app.pool, campaign_context, session, events_success).await
-        {
-            error!(&app.logger, "Analytics recording failed: {}", err; "campaign" => ?campaign_context, "err" => ?err)
-        }
-
-        Ok(true)
+        });
     }
 
-    /// This function calculates the fee for each validator.
+    /// This function calculates the fee for each validator and each `Event`.
     ///
     /// It then spends the given amounts for:
     ///
-    /// - `Publisher` - payout
-    ///
-    /// `Leader` and `Follower` - Validator fees
-    pub async fn spend_for_event(
-        pool: &DbPool,
-        campaign_remaining: &CampaignRemaining,
+    /// - [`Event`] `Publisher` - for payout
+    /// - `Leader` and `Follower` - for validator fees
+    pub async fn spend_for_events<C: Locked + 'static>(
```

Comment:
Although I haven't tested with multiple events per request, this should improve the performance as it's handling all events first and then checks and spends the budget of the campaign.